crates/fuzzing/src/oracles.rs

```diff
@@ -479,3 +476,189 @@ pub fn table_ops(config: crate::generators::Config, ops: crate::generators::tabl
+                    };
+                    match exp.kind {
+                        wasmparser::ExternalKind::Memory => {
+                            memory_name = Some(exp.field.to_string());
+                        }
+                        wasmparser::ExternalKind::Function => {
+                            func_name = Some(exp.field.to_string());
+                        }
+                        _ => {}
+                    }
+                    if memory_name.is_some() && func_name.is_some() {
+                        break;
+                    }
+                }
+            }
+            wasmparser::Payload::MemorySection(mut m) => {
+                for _ in 0..m.get_count() {
+                    let mem = match m.read() {
+                        Ok(mem) => mem,
+                        Err(_) => return None,
+                    };
+                    match mem {
+                        wasmparser::MemoryType::M32 { limits, .. } => {
+                            if limits.initial > 100 {
+                                return None;
+                            }
+                            if limits.maximum.is_none() || limits.maximum.unwrap() > 100 {
+                                return None;
+                            }
+                        }
+                        wasmparser::MemoryType::M64 { .. } => {
+                            // Disallow 64-bit memories altogether.
+                            return None;
+                        }
+                    }
+                }
+            }
+            _ => {}
+        }
+    }
+    if !memory_name.is_some() || !func_name.is_some() {
+        return None;
+    }
+
+    let memory_name = memory_name.unwrap();
+    let func_name = func_name.unwrap();
+    Some((memory_name, func_name))
+}
+
+/// Perform differential execution between Cranelift and wasmi, diffing the resulting memory image
+/// when execution terminates. This relies on the module-under-test to be instrumented to bound the
+/// execution time.
+pub fn differential_wasmi_execution(
+    module: &wasm_smith::Module,
+    config: &crate::generators::Config,
+) {
+    crate::init_fuzzing();
+
+    let wasm = module.to_bytes();
+
+    let (memory_name, func_name) = match differential_wasmi_accept_module(&wasm[..]) {
+        Some(pair) => pair,
+        None => return,
+    };
+
+    // Use wasmi initialization sequence to validate requirements.
+    let wasmi_module = match wasmi::Module::from_buffer(&wasm[..]) {
+        Ok(m) => m,
+        Err(_) => return,
+    };
+    if wasmi_module.deny_floating_point().is_err() {
+        return;
+    }
+    let wasmi_instance =
+        match wasmi::ModuleInstance::new(&wasmi_module, &wasmi::ImportsBuilder::default()) {
+            Ok(inst) => inst,
+            Err(_) => return,
+        };
+    if wasmi_instance.has_start() {
+        return;
+    }
+    let wasmi_instance = wasmi_instance.assert_no_start();
```

Comment:
This is mostly just for simplicity -- instantiate module first on both sides, then execute and compare. Also `start()` can't return a value, and it seemed better to (i) only execute one function, to economize on fuzzing input bytes (more efficient search) and (ii) allow a return-value comparison. Since we have `allow_start_export()` in the `Config` now I think this is probably fine, as we aren't rejecting any modules because of it?